src/main/java/org/wildfly/security/credential/source/CallbackHandlerCredentialSource.java

```diff
@@ -0,0 +1,78 @@
+/*
+ * JBoss, Home of Professional Open Source.
+ * Copyright 2016 Red Hat, Inc., and individual contributors
+ * as indicated by the @author tags.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.wildfly.security.credential.source;
+
+import java.io.IOException;
+import java.security.spec.AlgorithmParameterSpec;
+
+import javax.security.auth.callback.Callback;
+import javax.security.auth.callback.CallbackHandler;
+import javax.security.auth.callback.PasswordCallback;
+import javax.security.auth.callback.UnsupportedCallbackException;
+
+import org.wildfly.common.Assert;
+import org.wildfly.security.auth.SupportLevel;
+import org.wildfly.security.auth.callback.CredentialCallback;
+import org.wildfly.security.credential.Credential;
+import org.wildfly.security.credential.PasswordCredential;
+import org.wildfly.security.password.interfaces.ClearPassword;
+
+/**
+ * A credential source which is backed by a callback handler.
+ *
+ * @author <a href="mailto:david.lloyd@redhat.com">David M. Lloyd</a>
+ */
+public final class CallbackHandlerCredentialSource implements CredentialSource {
+    private final CallbackHandler callbackHandler;
+
+    /**
+     * Construct a new instance.
+     *
+     * @param callbackHandler the callback handler to use (must not be {@code null})
+     */
+    public CallbackHandlerCredentialSource(final CallbackHandler callbackHandler) {
+        Assert.checkNotNullParam("callbackHandler", callbackHandler);
+        this.callbackHandler = callbackHandler;
+    }
+
+    public SupportLevel getCredentialAcquireSupport(final Class<? extends Credential> credentialType, final String algorithmName, final AlgorithmParameterSpec parameterSpec) throws IOException {
+        return SupportLevel.POSSIBLY_SUPPORTED;
+    }
+
+    public <C extends Credential> C getCredential(final Class<C> credentialType, final String algorithmName, final AlgorithmParameterSpec parameterSpec) throws IOException {
+        if (PasswordCredential.class.isAssignableFrom(credentialType) && (algorithmName == null || algorithmName.equals(ClearPassword.ALGORITHM_CLEAR)) && parameterSpec == null) {
```

Comment:
Looking at both Peter and Pedro's PRs I was worried that handling these callbacks was starting to be duplicated as they both needed to solve the same problem for different reasons and in some cases handled subtly differently - so this should become the place we interact with the callbacks.